test w/ multiple ansible versions

kdelee opened the pull request:
first stab, some test failures -- probably valid ones

I believe this will end up running in zuul all as one job, we can poke at zuul config if we want to split it up

re: https://github.com/ansible/ansible-runner/issues/476

kdelee committed:
first stab, some test failures

I beleive this will end up running in zuul all as one job, we can poke at zuul config if we want to split it up

jladdjr commented on test/integration/test_events.py:
```diff
@@ -10,6 +10,16 @@
 
 
+@pytest.fixture(scope='session')
+def skipif_pre_ansible29():
```
based on the fixture's definition, looks like this method should actually be name `skipif_pre_ansible28`. Am I reading this correctly?

```python
if LooseVersion(pkg_resources.get_distribution('ansible').version) < LooseVersion('2.8'):
    pytest.skip(reason="Valid only on Ansible 2.8+")
```

jladdjr replied:
apart from that question, looks like a great / helpful fixture 💯

kdelee replied:
oops you are right

jladdjr commented on tox.ini:
```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist = linters, py3
+envlist = linters, ansible{27, 28, 29, -base}
```
slick - neat trick!

jladdjr reviewed (approved):
Had a question about the new `skipif_pre_ansible29` fixture, otherwise 💯 . Thanks for tackling this!

jladdjr commented:
Saw this pop up in github notifications and looked like a quick review. Heading back into PPL.. 😄

kdelee committed:
fix py36 and py37 tox envs

also fix name of fixture

kdelee committed:
fix ansible-base failure

ansible-base failure was just because of a deprecation warning
also fixed syntax error on calling pytest.skip for the 27 test

jladdjr commented on tox.ini:
```diff
@@ -30,7 +33,26 @@ commands=
+       ansible
+passenv = HOME
+skip_install = true
+commands=
```
Wondering if a few of these settings (`passenv`, `skip_install`, and `commands`) need to be here, or if they'll automatically be inherited because they're already in the `testenv` section.

kdelee commented:
recheck

kdelee commented:
https://github.com/ansible/project-config/pull/577 got merged a bit sooner than I suspected, so now we are running these checks here. ansible 2.8, 2.9 and devel are happy, but there are a number of problems with 2.7 that I could use some help interpreting.

The options seem like:
1) Fix tests and or runner to work w/ 2.7, leave check in
2) Disable check, but fix runner later and then renable
3) Drop support for 2.7, remove check

kdelee commented:
for reference, here are the ansible 2.7 errors on zuul on a fedora 31 vm:

```
=================================== FAILURES ===================================
________________________ test_cmdline_cmdline_override _________________________
[gw2] linux -- Python 3.7.7 /home/zuul/src/github.com/ansible/ansible-runner/.tox/ansible27/bin/python3

    def test_cmdline_cmdline_override():
        try:
            private_data_dir = tempfile.mkdtemp()
            play = [{'hosts': 'all', 'tasks': [{'debug': {'msg': random_string()}}]}]
    
            path = os.path.join(private_data_dir, 'project')
            os.makedirs(path)
    
            playbook = os.path.join(path, 'main.yaml')
            with open(playbook, 'w') as f:
                f.write(json.dumps(play))
            path = os.path.join(private_data_dir, 'inventory')
            os.makedirs(path)
    
            inventory = os.path.join(path, 'hosts')
            with open(inventory, 'w') as f:
                f.write('[all]\nlocalhost ansible_connection=local')
    
            # privateip: removed --hosts command line option from test beause it is
            # not a supported combination of cli options
            #cmdline('run', private_data_dir, '-p', playbook, '--hosts', 'all', '--cmdline', '-e foo=bar')
            cmdline('run', private_data_dir, '-p', playbook, '--cmdline', '-e foo=bar')
>           assert main() == 0
E           assert 2 == 0
E             -2
E             +0

test/integration/test___main__.py:241: AssertionError
----------------------------- Captured stdout call -----------------------------


PLAY [all] *********************************************************************


TASK [Gathering Facts] *********************************************************

fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "/bin/sh: /usr/bin/python: No such file or directory\n", "module_stdout": "", "msg": "The module failed to execute correctly, you probably need to set the interpreter.\nSee stdout/stderr for the exact error", "rc": 127}

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   


PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   

____________________________ test_cmdline_playbook _____________________________
[gw3] linux -- Python 3.7.7 /home/zuul/src/github.com/ansible/ansible-runner/.tox/ansible27/bin/python3

    def test_cmdline_playbook():
        try:
            private_data_dir = tempfile.mkdtemp()
            play = [{'hosts': 'all', 'tasks': [{'debug': {'msg': random_string()}}]}]
    
            path = os.path.join(private_data_dir, 'project')
            os.makedirs(path)
    
            playbook = os.path.join(path, 'main.yaml')
            with open(playbook, 'w') as f:
                f.write(json.dumps(play))
    
            path = os.path.join(private_data_dir, 'inventory')
            os.makedirs(path)
    
            inventory = os.path.join(path, 'hosts')
            with open(inventory, 'w') as f:
                f.write('[all]\nlocalhost ansible_connection=local')
    
            cmdline('run', private_data_dir, '-p', playbook, '--inventory', inventory)
    
>           assert main() == 0
E           assert 2 == 0
E             -2
E             +0

test/integration/test___main__.py:192: AssertionError
----------------------------- Captured stdout call -----------------------------


PLAY [all] *********************************************************************

TASK [Gathering Facts] *********************************************************

fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "/bin/sh: /usr/bin/python: No such file or directory\n", "module_stdout": "", "msg": "The module failed to execute correctly, you probably need to set the interpreter.\nSee stdout/stderr for the exact error", "rc": 127}

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   


PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   

______________ test_callback_plugin_saves_custom_stats[playbook0] ______________
[gw1] linux -- Python 3.7.7 /home/zuul/src/github.com/ansible/ansible-runner/.tox/ansible27/bin/python3

executor = <ansible_runner.runner.Runner object at 0x7f7af7afb4d0>
playbook = {'custom_set_stat.yml': '\n- name: custom set_stat calls should persist to the local disk so awx can save them\n  connection: local\n  hosts: all\n  tasks:\n    - set_stats:\n        data:\n          foo: "bar"\n'}

    @pytest.mark.parametrize('playbook', [
    {'custom_set_stat.yml': '''
    - name: custom set_stat calls should persist to the local disk so awx can save them
      connection: local
      hosts: all
      tasks:
        - set_stats:
            data:
              foo: "bar"
    '''},  # noqa
    ])
    def test_callback_plugin_saves_custom_stats(executor, playbook):
        executor.run()
        for event in executor.events:
            event_data = event.get('event_data', {})
            if 'artifact_data' in event_data:
>               assert event_data['artifact_data'] == {'foo': 'bar'}
E               AssertionError: assert {} == {'foo': 'bar'}
E                 Right contains more items:
E                 {'foo': 'bar'}
E                 Full diff:
E                 - {}
E                 + {'foo': 'bar'}

test/integration/test_display_callback.py:266: AssertionError
----------------------------- Captured stdout call -----------------------------


PLAY [custom set_stat calls should persist to the local disk so awx can save them] ***


TASK [Gathering Facts] *********************************************************

fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "/bin/sh: /usr/bin/python: No such file or directory\n", "module_stdout": "", "msg": "The module failed to execute correctly, you probably need to set the interpreter.\nSee stdout/stderr for the exact error", "rc": 127}

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   


PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   

________________ test_callback_plugin_task_args_leak[playbook0] ________________
[gw2] linux -- Python 3.7.7 /home/zuul/src/github.com/ansible/ansible-runner/.tox/ansible27/bin/python3

executor = <ansible_runner.runner.Runner object at 0x7f94f4e02990>
playbook = {'no_log_on_ok.yml': '\n- name: args should not be logged when no_log is set at the task or module level\n  connection...echo "PRIVATE"\n      no_log: true\n    - uri: url=https://example.org url_username="PUBLIC" url_password="PRIVATE"\n'}

    @pytest.mark.parametrize('playbook', [
    {'no_log_on_ok.yml': '''
    - name: args should not be logged when no_log is set at the task or module level
      connection: local
      hosts: all
      gather_facts: no
      tasks:
        - shell: echo "PUBLIC"
        - shell: echo "PRIVATE"
          no_log: true
        - uri: url=https://example.org url_username="PUBLIC" url_password="PRIVATE"
    '''},  # noqa
    ])
    def test_callback_plugin_task_args_leak(executor, playbook):
        executor.run()
        events = list(executor.events)
        assert events[0]['event'] == 'playbook_on_start'
        assert events[1]['event'] == 'playbook_on_play_start'
    
        # task 1
        assert events[2]['event'] == 'playbook_on_task_start'
>       assert events[3]['event'] == 'runner_on_start'
E       AssertionError: assert 'runner_on_failed' == 'runner_on_start'
E         - runner_on_failed
E         + runner_on_start

test/integration/test_display_callback.py:187: AssertionError
----------------------------- Captured stdout call -----------------------------


PLAY [args should not be logged when no_log is set at the task or module level] ***


TASK [shell] *******************************************************************

fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "/bin/sh: /usr/bin/python: No such file or directory\n", "module_stdout": "", "msg": "The module failed to execute correctly, you probably need to set the interpreter.\nSee stdout/stderr for the exact error", "rc": 127}

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   


PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   

_________ test_callback_plugin_censoring_does_not_overwrite[playbook0] _________
[gw3] linux -- Python 3.7.7 /home/zuul/src/github.com/ansible/ansible-runner/.tox/ansible27/bin/python3

executor = <ansible_runner.runner.Runner object at 0x7fb676843d90>
playbook = {'loop_with_no_log.yml': '\n- name: playbook variable should not be overwritten when using no log\n  connection: local...ms:\n        - "echo helloworld!"\n    - debug: msg="{{ command_register.results|map(attribute=\'stdout\')|list }}"\n'}

    @pytest.mark.parametrize('playbook', [
    {'loop_with_no_log.yml': '''
    - name: playbook variable should not be overwritten when using no log
      connection: local
      hosts: all
      gather_facts: no
      tasks:
        - command: "{{ item }}"
          register: command_register
          no_log: True
          with_items:
            - "echo helloworld!"
        - debug: msg="{{ command_register.results|map(attribute='stdout')|list }}"
    '''},  # noqa
    ])
    def test_callback_plugin_censoring_does_not_overwrite(executor, playbook):
        executor.run()
        events = list(executor.events)
        assert events[0]['event'] == 'playbook_on_start'
        assert events[1]['event'] == 'playbook_on_play_start'
    
        # task 1
        assert events[2]['event'] == 'playbook_on_task_start'
        # Ordering of task and item events may differ randomly
>       assert set(['runner_on_start', 'runner_item_on_ok', 'runner_on_ok']) == set([data['event'] for data in events[3:6]])
E       AssertionError: assert {'runner_item...ner_on_start'} == {'playbook_on_...er_on_failed'}
E         Extra items in the left set:
E         'runner_on_start'
E         'runner_item_on_ok'
E         'runner_on_ok'
E         Extra items in the right set:
E         'playbook_on_stats'
E         'runner_item_on_failed'...
E         
E         ...Full output truncated (5 lines hidden), use '-vv' to show

test/integration/test_display_callback.py:225: AssertionError
----------------------------- Captured stdout call -----------------------------

PLAY [playbook variable should not be overwritten when using no log] ***********


TASK [command] *****************************************************************

failed: [localhost] (item=None) => {"censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result", "changed": false}

fatal: [localhost]: FAILED! => {"censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result", "changed": false}

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   


PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   

____________ test_callback_plugin_records_notify_events[playbook0] _____________
[gw1] linux -- Python 3.7.7 /home/zuul/src/github.com/ansible/ansible-runner/.tox/ansible27/bin/python3

executor = <ansible_runner.runner.Runner object at 0x7f7af7be2ad0>
playbook = {'handle_playbook_on_notify.yml': '\n- name: handle playbook_on_notify events properly\n  connection: local\n  hosts: ...g="My Handler"\n  tasks:\n    - debug: msg="My Task"\n      changed_when: true\n      notify:\n        - my_handler\n'}

    @pytest.mark.parametrize('playbook', [
    {'handle_playbook_on_notify.yml': '''
    - name: handle playbook_on_notify events properly
      connection: local
      hosts: all
      handlers:
        - name: my_handler
          debug: msg="My Handler"
      tasks:
        - debug: msg="My Task"
          changed_when: true
          notify:
            - my_handler
    '''},  # noqa
    ])
    @pytest.mark.skipif(ANSIBLE_VERSION < '2.5', reason="v2_playbook_on_notify doesn't work before ansible 2.5")
    def test_callback_plugin_records_notify_events(executor, playbook):
        executor.run()
        assert len(list(executor.events))
        notify_events = [x for x in executor.events if x['event'] == 'playbook_on_notify']
>       assert len(notify_events) == 1
E       assert 0 == 1
E         -0
E         +1

test/integration/test_display_callback.py:292: AssertionError
----------------------------- Captured stdout call -----------------------------


PLAY [handle playbook_on_notify events properly] *******************************

TASK [Gathering Facts] *********************************************************

fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "/bin/sh: /usr/bin/python: No such file or directory\n", "module_stdout": "", "msg": "The module failed to execute correctly, you probably need to set the interpreter.\nSee stdout/stderr for the exact error", "rc": 127}

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   


PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   

_____________________ test_module_level_no_log[playbook0] ______________________
[gw1] linux -- Python 3.7.7 /home/zuul/src/github.com/ansible/ansible-runner/.tox/ansible27/bin/python3

executor = <ansible_runner.runner.Runner object at 0x7f7af78eb6d0>
playbook = {'no_log_module_with_var.yml': '\n- name: ensure that module-level secrets are redacted\n  connection: local\n  hosts:...  url: https://example.org\n        url_username: john-jacob-jingleheimer-schmidt\n        url_password: "{{ pw }}"\n'}

    @pytest.mark.parametrize('playbook', [
    {'no_log_module_with_var.yml': '''
    - name: ensure that module-level secrets are redacted
      connection: local
      hosts: all
      vars:
        - pw: SENSITIVE
      tasks:
        - uri:
            url: https://example.org
            url_username: john-jacob-jingleheimer-schmidt
            url_password: "{{ pw }}"
    '''},  # noqa
    ])
    def test_module_level_no_log(executor, playbook):
        # It's possible for `no_log=True` to be defined at the _module_ level,
        # e.g., for the URI module password parameter
        # This test ensures that we properly redact those
        executor.run()
        assert len(list(executor.events))
>       assert 'john-jacob-jingleheimer-schmidt' in json.dumps(list(executor.events))
E       assert 'john-jacob-jingleheimer-schmidt' in '[{"uuid": "123bf5da-54f3-4b65-8e81-e9b9579e95d9", "counter": 1, "stdout": "", "start_line": 0, "end_line": 0, "runner... {"localhost": 1}, "rescued": 0, "skipped": {}, "artifact_data": {}, "uuid": "030b8e35-2894-4ea5-bc7b-49574c296a2c"}}]'
E        +  where '[{"uuid": "123bf5da-54f3-4b65-8e81-e9b9579e95d9", "counter": 1, "stdout": "", "start_line": 0, "end_line": 0, "runner... {"localhost": 1}, "rescued": 0, "skipped": {}, "artifact_data": {}, "uuid": "030b8e35-2894-4ea5-bc7b-49574c296a2c"}}]' = <function dumps at 0x7f7afc5a5ef0>([{'counter': 1, 'created': '2020-07-31T21:00:51.576003', 'end_line': 0, 'event': 'playbook_on_start', ...}, {'counter'...iled', ...}, {'counter': 5, 'created': '2020-07-31T21:00:55.650446', 'end_line': 9, 'event': 'playbook_on_stats', ...}])
E        +    where <function dumps at 0x7f7afc5a5ef0> = json.dumps
E        +    and   [{'counter': 1, 'created': '2020-07-31T21:00:51.576003', 'end_line': 0, 'event': 'playbook_on_start', ...}, {'counter'...iled', ...}, {'counter': 5, 'created': '2020-07-31T21:00:55.650446', 'end_line': 9, 'event': 'playbook_on_stats', ...}] = list(<generator object Runner.events at 0x7f7af7b8d9d0>)
E        +      where <generator object Runner.events at 0x7f7af7b8d9d0> = <ansible_runner.runner.Runner object at 0x7f7af78eb6d0>.events

test/integration/test_display_callback.py:318: AssertionError
----------------------------- Captured stdout call -----------------------------


PLAY [ensure that module-level secrets are redacted] ***************************

TASK [Gathering Facts] *********************************************************
fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "/bin/sh: /usr/bin/python: No such file or directory\n", "module_stdout": "", "msg": "The module failed to execute correctly, you probably need to set the interpreter.\nSee stdout/stderr for the exact error", "rc": 127}

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   


PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   

________________________________ test_role_run _________________________________
[gw2] linux -- Python 3.7.7 /home/zuul/src/github.com/ansible/ansible-runner/.tox/ansible27/bin/python3

    def test_role_run():
        rc = main(['-r', 'benthomasson.hello_role',
                   '--hosts', 'localhost',
                   '--roles-path', 'test/integration/roles',
                   'run',
                   "test/integration"])
>       assert rc == 0
E       assert 2 == 0
E         -2
E         +0

test/integration/test_main.py:123: AssertionError
----------------------------- Captured stdout call -----------------------------


PLAY [localhost] ***************************************************************


TASK [Gathering Facts] *********************************************************

fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "/bin/sh: /usr/bin/python: No such file or directory\n", "module_stdout": "", "msg": "The module failed to execute correctly, you probably need to set the interpreter.\nSee stdout/stderr for the exact error", "rc": 127}

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   


PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   

______________________________ test_role_logfile _______________________________
[gw2] linux -- Python 3.7.7 /home/zuul/src/github.com/ansible/ansible-runner/.tox/ansible27/bin/python3

    def test_role_logfile():
        try:
            rc = main(['-r', 'benthomasson.hello_role',
                       '--hosts', 'localhost',
                       '--roles-path', 'test/integration/project/roles',
                       '--logfile', 'new_logfile',
                       'run',
                       'test/integration'])
            assert os.path.exists('new_logfile')
>           assert rc == 0
E           assert 2 == 0
E             -2
E             +0

test/integration/test_main.py:146: AssertionError
----------------------------- Captured stdout call -----------------------------


PLAY [localhost] ***************************************************************


TASK [Gathering Facts] *********************************************************

fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "/bin/sh: /usr/bin/python: No such file or directory\n", "module_stdout": "", "msg": "The module failed to execute correctly, you probably need to set the interpreter.\nSee stdout/stderr for the exact error", "rc": 127}

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   


PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   

_________________________ test_role_run_artifacts_dir __________________________
[gw2] linux -- Python 3.7.7 /home/zuul/src/github.com/ansible/ansible-runner/.tox/ansible27/bin/python3

    def test_role_run_artifacts_dir():
        rc = main(['-r', 'benthomasson.hello_role',
                   '--hosts', 'localhost',
                   '--roles-path', 'test/integration/roles',
                   '--artifact-dir', 'otherartifacts',
                   'run',
                   "test/integration"])
>       assert rc == 0
E       assert 2 == 0
E         -2
E         +0

test/integration/test_main.py:213: AssertionError
----------------------------- Captured stdout call -----------------------------

PLAY [localhost] ***************************************************************


TASK [Gathering Facts] *********************************************************

fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "/bin/sh: /usr/bin/python: No such file or directory\n", "module_stdout": "", "msg": "The module failed to execute correctly, you probably need to set the interpreter.\nSee stdout/stderr for the exact error", "rc": 127}

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   


PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   

___________________________ test_role_run_inventory ____________________________
[gw2] linux -- Python 3.7.7 /home/zuul/src/github.com/ansible/ansible-runner/.tox/ansible27/bin/python3

    def test_role_run_inventory():
    
        with temp_directory() as temp_dir:
            ensure_directory(os.path.join(temp_dir, 'inventory'))
            shutil.copy(os.path.join(HERE, 'inventory/localhost'), os.path.join(temp_dir, 'inventory/localhost'))
    
            rc = main(['-r', 'benthomasson.hello_role',
                       '--hosts', 'localhost',
                       '--roles-path', os.path.join(HERE, 'project/roles'),
                       '--inventory', 'localhost',
                       'run',
                       temp_dir])
>       assert rc == 0
E       assert 2 == 0
E         -2
E         +0

test/integration/test_main.py:279: AssertionError
----------------------------- Captured stdout call -----------------------------
/tmp/tmp7pm2s1pr

PLAY [localhost] ***************************************************************

TASK [Gathering Facts] *********************************************************
fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "/bin/sh: /usr/bin/python: No such file or directory\n", "module_stdout": "", "msg": "The module failed to execute correctly, you probably need to set the interpreter.\nSee stdout/stderr for the exact error", "rc": 127}

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   


PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   

=============================== warnings summary ===============================
```

kdelee committed:
Skip tests incompatible with 2.7 and earlier

Not a lot of energy around making these 2.7 compat as is an older
ansible, not going to block this work any further waiting to see what we
might want to do around these.

If it becomes a priority to add more coverage for 2.7 in future we can
start by looking at these tests.

kdelee committed:
workaround ansibl27 interpreter discovery

kdelee commented:
:exclamation: finally got ansible 2.7 differences ironed out.

as I said in commit message:

> Not a lot of energy around making these 2.7 compat as is an older
ansible, not going to block this work any further waiting to see what we
might want to do around these.

> If it becomes a priority to add more coverage for 2.7 in future we can
start by looking at these tests.

Is not a perfect solution, but we have some coverage which is better than none. If it becomes higher priority we can go back over the tests we are skipping and figure out what to do.

Going to put through gate check now
